File: ruby/saddle-points/saddle_points.rb
class Matrix

  def initialize(matrix_text)
    @matrix = matrix_text.split("\n").map { |e| e.split(' ').map { |k| k.to_i } }
  end

  def rows
    @matrix
  end

  def columns
    @matrix[0].map.with_index { |e, i| @matrix.map { |r| r[i] } }
  end

  def saddle_points
    max_index_per_row = []
    min_index_per_col = []
    rows.each_with_index { |row, i| row.map.with_index { |c, k| c == row.max ? k : nil }.compact.each { |m| max_index_per_row << [i, m] } }
    columns.each_with_index { |col, i| col.map.with_index { |c, k| c == col.min ? k : nil }.compact.each { |m| min_index_per_col << [m, i] } }

    max_index_per_row & min_index_per_col

  end

end

File: python/saddle-points/saddle_points.py
def saddle_points(matrix):
    if len(matrix) == 0 or len(matrix[0]) == 0:
        return [{}]

    saddle_points = [] 
    try:
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                if matrix[i][j] == max(matrix[i]) and matrix[i][j] == min((e[j] for e in matrix)):
                    saddle_points.append({'row': i + 1, 'column': j + 1})
    except IndexError:
        raise ValueError('Irregular matrix')
    
    return saddle_points if len(saddle_points) > 0 else [{}]

